src/abi/anchor.rs

```diff
@@ -17,3 +29,343 @@ pub fn discriminator(namespace: &'static str, name: &str) -> Vec<u8> {
+    }
+}
+
+/// Generate the IDL instructions for a contract.
+fn idl_instructions(
+    contract: &Contract,
+    type_manager: &mut TypeManager,
+    ns: &Namespace,
+) -> Vec<IdlInstruction> {
+    let mut instructions: Vec<IdlInstruction> = Vec::new();
+
+    if !contract.have_constructor(ns) {
+        instructions.push(IdlInstruction {
+            name: "new".to_string(),
+            docs: None,
+            accounts: vec![IdlAccountItem::IdlAccount(IdlAccount {
+                name: "data_account".to_string(),
+                is_mut: true,
+                is_signer: false,
+                is_optional: Some(false),
+                docs: None,
+                pda: None,
+                relations: vec![],
+            })],
+            args: vec![],
+            returns: None,
+        })
+    }
+
+    for func_no in contract.all_functions.keys() {
+        if !ns.functions[*func_no].is_public() {
+            continue;
+        }
+
+        let func = &ns.functions[*func_no];
+        let tags = idl_docs(func);
+
+        let accounts = match &func.mutability {
+            Mutability::Pure(_) => {
+                vec![]
+            }
+            Mutability::View(_) => {
+                vec![IdlAccountItem::IdlAccount(IdlAccount {
+                    name: "data_account".to_string(),
+                    is_mut: false,
+                    is_signer: false,
+                    is_optional: Some(false),
+                    docs: None,
+                    pda: None,
+                    relations: vec![],
+                })]
+            }
+            _ => {
+                vec![IdlAccountItem::IdlAccount(IdlAccount {
+                    name: "data_account".to_string(),
+                    is_mut: true,
+                    is_signer: false,
+                    is_optional: Some(false),
+                    docs: None,
+                    pda: None,
+                    relations: vec![],
+                })]
+            }
+        };
+
+        let mut args: Vec<IdlField> = Vec::with_capacity(func.params.len());
+        for (item_no, item) in func.params.iter().enumerate() {
+            let name = if item.id.is_none() || item.id.as_ref().unwrap().name.is_empty() {
+                format!("arg_{}", item_no)
+            } else {
+                item.id.as_ref().unwrap().name.clone()
+            };
+            args.push(IdlField {
+                name,
+                docs: None,
+                ty: type_manager.convert(&item.ty),
+            });
+        }
+
+        let returns = if func.returns.is_empty() {
+            None
+        } else if func.returns.len() == 1 {
+            Some(type_manager.convert(&func.returns[0].ty))
+        } else {
+            Some(type_manager.build_struct_for_return(func))
+        };
+
+        instructions.push(IdlInstruction {
+            name: if func.is_constructor() {
+                "new".to_string()
+            } else {
+                func.name.clone()
```

Comment:
Might use mangled name for overloaded functions? I see later down in the tests you use `.instructions.iter().find(|item| item.name == name)` to match against a function

Reply:
Fixed